File: typescript/src/sandbox.ts
import type { Connection } from './connection.js';
import {
  CONNECT_SANDBOX_SESSION_PATH_SESSION_ID,
  CONNECT_SANDBOX_SESSION_QUERY_TICKET,
  resolveOpenAPIOperation,
} from './generated/openapi.js';
import type {
  CreateSandboxSessionRequest,
  CreateSandboxSessionResponse,
  Sandbox,
  SandboxCreateRequest,
  SandboxSessionKind,
  SandboxWorkspace,
} from './types.js';

/**
 * Bound on a single frame through the session relay, in either direction. A
 * peer writing more has its session closed rather than forwarded, so chunk
 * writes at or below this.
 */
export const SANDBOX_RELAY_MAX_FRAME_BYTES = 1 << 20;

/**
 * Exec exit codes arrive as the WebSocket close status, `4000 + code`. A close
 * outside this range is a transport failure, not a process exit.
 */
export const SANDBOX_EXEC_CLOSE_STATUS_BASE = 4000;
export const SANDBOX_EXEC_CLOSE_STATUS_MAX = SANDBOX_EXEC_CLOSE_STATUS_BASE + 255;

/** Decodes a process exit code from a WebSocket close status. */
export function sandboxExecExitCode(closeStatus: number): number | undefined {
  if (closeStatus < SANDBOX_EXEC_CLOSE_STATUS_BASE || closeStatus > SANDBOX_EXEC_CLOSE_STATUS_MAX) {
    return undefined;
  }
  return closeStatus - SANDBOX_EXEC_CLOSE_STATUS_BASE;
}

export interface SandboxWaitOptions {
  /** Defaults to 120_000 ms. */
  timeoutMs?: number;
  /** Defaults to 1_000 ms. */
  pollIntervalMs?: number;
  signal?: AbortSignal;
}

export interface SandboxWebSocketConnectOptions {
  /** Connection-level headers, including the management Authorization header. */
  readonly headers: Readonly<Record<string, string>>;
}

/** Creates an authenticated WebSocket for a sandbox relay connection. */
export type SandboxWebSocketFactory = (
  url: string,
  options: SandboxWebSocketConnectOptions,
) => WebSocket;

export interface SandboxSessionOptions {
  command?: string[];
  /**
   * PTY size, defaulting to 24x80. Fixed for the session's life — the relay
   * has no resize channel, so a terminal resized mid-session cannot tell the
   * sandbox.
   */
  rows?: number;
  columns?: number;
  /**
   * Overrides where the WebSocket dials, defaulting to the connection's base
   * URL. Needed when the API is reached through a proxy that doesn't forward
   * WebSocket upgrades — the relay is deliberately not proxied.
   */
  relayBaseUrl?: string;
  /**
   * Overrides WebSocket construction. The factory receives every configured
   * connection header, including the management Authorization credential.
   */
  webSocketFactory?: SandboxWebSocketFactory;
  /**
   * Legacy constructor override. Header-capable Node implementations receive
   * the connection headers in their constructor options. Prefer
   * `webSocketFactory` for new integrations.
   */
  webSocketImpl?: typeof WebSocket;
}

/**
 * A live sandbox session: a raw bidirectional byte stream plus the process
 * exit code once it closes.
 *
 * For `pty` this is terminal traffic; for `exec` it is the process's stdout
 * and stderr **interleaved on one stream** — the relay multiplexes both, so
 * they cannot be separated client-side. Message boundaries are otherwise
 * opaque, except that `closeInput` sends the reserved zero-length binary
 * stdin-EOF message.
 */
export class SandboxSession {
  private readonly closed: Promise<number | undefined>;
  private inputClosed = false;

  constructor(private readonly socket: WebSocket) {
    this.closed = new Promise((resolve) => {
      socket.addEventListener('close', (event) => resolve(sandboxExecExitCode(event.code)), { once: true });
    });
  }

  /** Sends bytes to the sandbox's stdin. */
  send(data: Uint8Array | string): void {
    if (this.inputClosed) {
      throw new Error('postgrip-agent: sandbox session input is closed');
    }
    const bytes = sandboxInputBytes(data);
    // An empty binary message is the stdin-EOF control. Preserve send's byte
    // stream semantics by making empty writes no-ops; closeInput owns it.
    if (bytes.byteLength === 0) return;
    this.socket.send(bytes);
  }

  /** Signals stdin EOF while keeping output and exit-status delivery open. */
  closeInput(): void {
    if (this.inputClosed) return;
    this.socket.send(new Uint8Array(0));
    this.inputClosed = true;
  }

  /** Registers a handler for sandbox output. */
  onData(handler: (chunk: Uint8Array) => void): void {
    this.socket.addEventListener('message', (event: MessageEvent) => {
      const data = event.data;
      if (data instanceof ArrayBuffer) handler(new Uint8Array(data));
      else if (data instanceof Uint8Array) handler(data);
      else if (typeof data === 'string') handler(new TextEncoder().encode(data));
    });
  }

  /**
   * Resolves with the process exit code when the session ends.
   *
   * Resolves `undefined` when the close status carried no exit code — which
   * means the transport ended rather than the process, and must not be read as
   * a successful run.
   */
  async exitCode(): Promise<number | undefined> {
    return this.closed;
  }

  /** Closes the session immediately. */
  close(): void {
    this.socket.close();
  }
}

export interface SandboxExecOptions extends Omit<SandboxSessionOptions, 'command'> {
  /** Bytes delivered to stdin before the SDK signals EOF. Defaults to empty. */
  stdin?: Uint8Array | string;
}

/**
 * Sandbox lifecycle and execution.
 *
 * Sandbox endpoints authenticate on the management lane: construct the
 * Connection with `authToken` set to a management token. An agent access token
 * is rejected by every endpoint here.
 */
export class SandboxClient {
  constructor(private readonly connection: Connection) {}

  async create(request: SandboxCreateRequest): Promise<Sandbox> {
    return this.connection.createSandbox(request);
  }

  async list(): Promise<Sandbox[]> {
    return this.connection.listSandboxes();
  }

  async get(sandboxId: string, signal?: AbortSignal): Promise<Sandbox> {
    return this.connection.getSandbox(requireSandboxId(sandboxId), signal);
  }

  async start(sandboxId: string): Promise<Sandbox> {
    return this.connection.startSandbox(requireSandboxId(sandboxId));
  }

  async stop(sandboxId: string): Promise<Sandbox> {
    return this.connection.stopSandbox(requireSandboxId(sandboxId));
  }

  async delete(sandboxId: string): Promise<Sandbox> {
    return this.connection.deleteSandbox(requireSandboxId(sandboxId));
  }

  async uploadWorkspace(
    archive: BodyInit,
    metadata: { repositoryName?: string; revision?: string } = {},
  ): Promise<SandboxWorkspace> {
    return this.connection.uploadWorkspace(archive, metadata);
  }

  async createSession(
    sandboxId: string,
    request: CreateSandboxSessionRequest = {},
  ): Promise<CreateSandboxSessionResponse> {
    return this.connection.createSandboxSession(requireSandboxId(sandboxId), request);
  }

  /**
   * Polls until the sandbox is ready, fails, or the wait expires.
   *
   * Readiness is `observedState === 'running'` **and**
   * `observedGeneration >= generation`. State alone can predate a start or
   * stop the assigned agent hasn't observed yet, so checking only the state
   * can return a sandbox that is about to stop.
   */
  async waitUntilRunning(sandboxId: string, options: SandboxWaitOptions = {}): Promise<Sandbox> {
    const timeoutMs = options.timeoutMs ?? 120_000;
    const pollIntervalMs = options.pollIntervalMs ?? 1_000;
    const deadline = Date.now() + timeoutMs;
    let last: Sandbox | undefined;

    for (;;) {
      options.signal?.throwIfAborted();
      // Bound the request itself, not just the gaps between requests. Without
      // a signal a stalled management call stays pending forever, and the loop
      // never reaches the deadline check below — so waitUntilRunning could
      // outlive its own timeout indefinitely. The signal combines the caller's
      // with what remains of the deadline, so whichever comes first wins.
      const remaining = deadline - Date.now();
      const pollSignal = combineAbortSignals(options.signal, remaining > 0 ? remaining : 1);
      const record = await this.get(sandboxId, pollSignal);
      last = record;
      // Both readiness and failure are read at the current generation. A
      // `failed` left over from the previous generation does not describe the
      // state just requested, so treating it as terminal made a start issued
      // against a failed sandbox reject during the reconciliation window —
      // exactly when a caller is trying to recover.
      const observed = (record.observedGeneration ?? 0) >= (record.generation ?? 0);
      if (record.observedState === 'running' && observed) {
        return record;
      }
      if (record.observedState === 'failed' && observed) {
        throw new Error(
          `postgrip-agent: sandbox ${sandboxId} failed: ` +
            (record.failureMessage || record.failureCode || 'no failure detail reported'),
        );
      }
      if (Date.now() >= deadline) {
        throw new Error(
          `postgrip-agent: sandbox ${sandboxId} was not running within ${timeoutMs}ms ` +
            `(last observed state: ${last?.observedState ?? 'unknown'})`,
        );
      }
      await new Promise((resolve) => setTimeout(resolve, pollIntervalMs));
    }
  }

  /**
   * Creates a session and dials the relay.
   *
   * The sandbox must already be running; while it comes up the server rejects
   * session creation with a retryable 400, so call `waitUntilRunning` first.
   * The ticket is single-use and short-lived, which is why creating and
   * dialling are one call.
   */
  async openSession(
    sandboxId: string,
    kind: SandboxSessionKind,
    options: SandboxSessionOptions = {},
  ): Promise<SandboxSession> {
    if (kind === 'exec' && (!options.command || options.command.length === 0)) {
      throw new Error('postgrip-agent: sandbox exec requires a command');
    }
    const relayHeaders: Record<string, string> = {};
    this.connection.managementHeaders().forEach((value, name) => {
      relayHeaders[name] = value;
    });
    Object.freeze(relayHeaders);
    // Resolve the transport before creating the single-use server session. A
    // missing implementation must not burn a ticket or start an exec command.
    const createWebSocket = await resolveSandboxWebSocketFactory(options, relayHeaders);
    const session = await this.createSession(sandboxId, {
      kind,
      command: options.command,
      rows: options.rows,
      columns: options.columns,
    });
    const url = sandboxRelayUrl(
      options.relayBaseUrl ?? this.connection.baseUrl,
      session.id!,
      session.ticket!,
    );
    // The ticket selects and authorizes the session; the management credential
    // separately authenticates the request. Both are required by the contract.
    const socket = createWebSocket(url, { headers: relayHeaders });
    socket.binaryType = 'arraybuffer';
    await new Promise<void>((resolve, reject) => {
      socket.addEventListener('open', () => resolve(), { once: true });
      socket.addEventListener('error', () => reject(new Error('postgrip-agent: sandbox relay dial failed')), {
        once: true,
      });
    });
    return new SandboxSession(socket);
  }

  /**
   * Runs a command in the sandbox and resolves with its exit code and output.
   *
   * `output` carries stdout and stderr interleaved — the relay is one stream.
   */
  async exec(
    sandboxId: string,
    command: string[],
    options: SandboxExecOptions = {},
  ): Promise<{ exitCode: number | undefined; output: Uint8Array }> {
    const { stdin, ...sessionOptions } = options;
    // Validate deterministic local failures before createSession starts a
    // remote command and consumes its single-use relay ticket.
    const stdinBytes = stdin === undefined ? undefined : sandboxInputBytes(stdin);
    const session = await this.openSession(sandboxId, 'exec', { ...sessionOptions, command });
    const chunks: Uint8Array[] = [];
    session.onData((chunk) => chunks.push(chunk));
    try {
      if (stdinBytes !== undefined) session.send(stdinBytes);
      session.closeInput();
    } catch (error) {
      // A socket can fail after the remote exec has started. The caller has no
      // session handle from this convenience method, so tear it down here.
      try {
        session.close();
      } catch {
        // Preserve the stdin setup failure; close is best-effort cleanup.
      }
      throw error;
    }
    const exitCode = await session.exitCode();
    let total = 0;
    for (const c of chunks) total += c.byteLength;
    const output = new Uint8Array(total);
    let offset = 0;
    for (const c of chunks) {
      output.set(c, offset);
      offset += c.byteLength;
    }
    return { exitCode, output };
  }
}

function sandboxInputBytes(data: Uint8Array | string): Uint8Array {
  const bytes = typeof data === 'string' ? new TextEncoder().encode(data) : data;
  if (bytes.byteLength > SANDBOX_RELAY_MAX_FRAME_BYTES) {
    throw new Error(
      `postgrip-agent: sandbox write of ${bytes.byteLength} bytes exceeds the relay frame limit ` +
        `(${SANDBOX_RELAY_MAX_FRAME_BYTES}); chunk your writes`,
    );
  }
  return bytes;
}

type HeaderCapableWebSocketConstructor = new (
  url: string,
  protocols?: string | string[],
  options?: { headers?: Readonly<Record<string, string>> },
) => WebSocket;

async function resolveSandboxWebSocketFactory(
  options: SandboxSessionOptions,
  headers: Readonly<Record<string, string>>,
): Promise<SandboxWebSocketFactory> {
  if (options.webSocketFactory) {
    return options.webSocketFactory;
  }
  if (options.webSocketImpl) {
    const WebSocketImpl = options.webSocketImpl as unknown as HeaderCapableWebSocketConstructor;
    return (url, connectOptions) => new WebSocketImpl(url, [], { headers: connectOptions.headers });
  }

  // The standard WebSocket API cannot attach an Authorization header. Use the
  // header-capable `ws` implementation on Node/Bun even when a global
  // WebSocket exists (Node 22+), otherwise token-authenticated relays fail.
  if (typeof process !== 'undefined' && process.versions?.node) {
    const { WebSocket: NodeWebSocket } = await import('ws');
    return (url, connectOptions) => new NodeWebSocket(url, {
      headers: { ...connectOptions.headers },
    }) as unknown as WebSocket;
  }

  const WebSocketImpl = globalThis.WebSocket;
  if (!WebSocketImpl) {
    throw new Error(
      'postgrip-agent: no WebSocket implementation available; pass webSocketFactory',
    );
  }
  if (Object.keys(headers).length > 0) {
    throw new Error(
      'postgrip-agent: browser WebSocket cannot send configured relay headers; ' +
        'authenticate with same-origin cookies or pass webSocketFactory in a header-capable runtime',
    );
  }
  return (url) => new WebSocketImpl(url);
}

/** Builds the ws(s):// relay URL from an http(s) API base URL. */
export function sandboxRelayUrl(baseUrl: string, sessionId: string, ticket: string): string {
  const base = baseUrl.trim().replace(/\/+$/, '');
  let origin: string;
  if (base.startsWith('https://')) origin = `wss://${base.slice('https://'.length)}`;
  else if (base.startsWith('http://')) origin = `ws://${base.slice('http://'.length)}`;
  else if (base.startsWith('wss://') || base.startsWith('ws://')) origin = base;
  else throw new Error(`postgrip-agent: sandbox relay base must be http(s) or ws(s): ${baseUrl}`);
  const operation = resolveOpenAPIOperation(
    'connectSandboxSession',
    { [CONNECT_SANDBOX_SESSION_PATH_SESSION_ID]: sessionId },
    new URLSearchParams({ [CONNECT_SANDBOX_SESSION_QUERY_TICKET]: ticket }),
  );
  return `${origin}${operation.path}`;
}

function requireSandboxId(sandboxId: string): string {
  if (!sandboxId) {
    // Otherwise this builds /api/v1/sandboxes/ and hits the collection route.
    throw new Error('postgrip-agent: sandbox id is required');
  }
  return sandboxId;
}

/**
 * A signal that aborts on the caller's signal or after `timeoutMs`, whichever
 * comes first.
 *
 * `AbortSignal.any` and `AbortSignal.timeout` are both Node 20+ / modern
 * browsers, which this package already targets, but they are guarded anyway:
 * on a runtime without them, falling back to the caller's signal alone leaves
 * the previous behaviour rather than throwing at an unrelated call site.
 */
function combineAbortSignals(signal: AbortSignal | undefined, timeoutMs: number): AbortSignal | undefined {
  if (typeof AbortSignal?.timeout !== 'function') {
    return signal;
  }
  const deadlineSignal = AbortSignal.timeout(timeoutMs);
  if (!signal) {
    return deadlineSignal;
  }
  if (typeof AbortSignal.any !== 'function') {
    return signal;
  }
  return AbortSignal.any([signal, deadlineSignal]);
}

File: src/postgrip_agent/sandbox.py
"""Sandbox platform client.

Sandboxes are persistent development environments assigned to one of your
agents. These endpoints authenticate on the *management* lane — construct the
Connection with ``auth_token`` set to a management token. An agent access token
is rejected by every endpoint here.
"""

from __future__ import annotations

import time
from typing import IO, TYPE_CHECKING, Any, Iterable
from urllib.error import HTTPError
from urllib.parse import quote, urlsplit
from urllib.request import Request, urlopen

if TYPE_CHECKING:  # pragma: no cover - import cycle only matters for typing
    from .client import Connection

#: Bound on a single frame through the session relay, in either direction. A
#: peer writing more has its session closed rather than forwarded, so chunk
#: writes at or below this.
SANDBOX_RELAY_MAX_FRAME_BYTES = 1 << 20

#: Exec exit codes arrive as the WebSocket close status, ``4000 + code``. A
#: close outside this range is a transport failure, not a process exit.
SANDBOX_EXEC_CLOSE_STATUS_BASE = 4000
SANDBOX_EXEC_CLOSE_STATUS_MAX = SANDBOX_EXEC_CLOSE_STATUS_BASE + 255

#: Server cap on a workspace archive upload; larger uploads are rejected 413.
SANDBOX_WORKSPACE_MAX_UPLOAD_BYTES = 512 << 20


def sandbox_exec_exit_code(close_status: int) -> int | None:
    """Decode a process exit code from a WebSocket close status.

    Returns ``None`` when the status carried no exit code — that means the
    transport ended rather than the process, and must not be read as success.
    """
    if close_status < SANDBOX_EXEC_CLOSE_STATUS_BASE or close_status > SANDBOX_EXEC_CLOSE_STATUS_MAX:
        return None
    return close_status - SANDBOX_EXEC_CLOSE_STATUS_BASE


def sandbox_relay_url(base_url: str, session_id: str, ticket: str) -> str:
    """Build the ``ws(s)://`` relay URL from an ``http(s)`` API base URL."""
    base = base_url.strip().rstrip("/")
    if base.startswith("https://"):
        origin = "wss://" + base[len("https://"):]
    elif base.startswith("http://"):
        origin = "ws://" + base[len("http://"):]
    elif base.startswith(("wss://", "ws://")):
        origin = base
    else:
        raise ValueError(f"postgrip-agent: sandbox relay base must be http(s) or ws(s): {base_url}")
    return (
        f"{origin}/api/v1/sandbox-sessions/{quote(session_id, safe='')}"
        f"/connect?ticket={quote(ticket, safe='')}"
    )


def sandbox_is_ready(record: dict[str, Any]) -> bool:
    """Readiness is observed state AND generation.

    A ``running`` reading can predate a start or stop the assigned agent has
    not observed yet, so state alone can describe a sandbox that is about to
    change again.
    """
    return (
        record.get("observedState") == "running"
        and int(record.get("observedGeneration") or 0) >= int(record.get("generation") or 0)
    )


class SandboxClient:
    """Sandbox lifecycle and execution."""

    def __init__(self, connection: "Connection"):
        self.connection = connection

    # --- lifecycle ---------------------------------------------------------

    def create(self, request: dict[str, Any]) -> dict[str, Any]:
        """Provision a sandbox. Returns once the record exists — the sandbox is
        not yet running, so follow with :meth:`wait_until_running`."""
        return self.connection.request("POST", "/api/v1/sandboxes", request)

    def list(self) -> list[dict[str, Any]]:
        """List live sandboxes.

        The endpoint returns an envelope, not a bare array; decoding it as an
        array would silently yield nothing.
        """
        response = self.connection.request("GET", "/api/v1/sandboxes") or {}
        return response.get("sandboxes") or []

    def get(self, sandbox_id: str) -> dict[str, Any]:
        return self.connection.request("GET", _sandbox_path(sandbox_id))

    def start(self, sandbox_id: str) -> dict[str, Any]:
        return self.connection.request("POST", _sandbox_path(sandbox_id) + "/start")

    def stop(self, sandbox_id: str) -> dict[str, Any]:
        return self.connection.request("POST", _sandbox_path(sandbox_id) + "/stop")

    def delete(self, sandbox_id: str) -> dict[str, Any]:
        return self.connection.request("DELETE", _sandbox_path(sandbox_id))

    def create_session(
        self,
        sandbox_id: str,
        *,
        kind: str = "pty",
        command: list[str] | None = None,
        rows: int | None = None,
        columns: int | None = None,
    ) -> dict[str, Any]:
        """Mint a single-use relay ticket.

        The sandbox must already be running; while it comes up the server
        returns a retryable 400, so call :meth:`wait_until_running` first.
        """
        if kind not in ("pty", "exec"):
            raise ValueError("postgrip-agent: sandbox session kind must be 'pty' or 'exec'")
        if kind == "exec" and not command:
            raise ValueError("postgrip-agent: sandbox exec requires a command")
        body: dict[str, Any] = {"kind": kind}
        if command is not None:
            body["command"] = command
        if rows is not None:
            body["rows"] = rows
        if columns is not None:
            body["columns"] = columns
        return self.connection.request("POST", _sandbox_path(sandbox_id) + "/sessions", body)

    def upload_workspace(
        self,
        archive: bytes | IO[bytes],
        *,
        repository_name: str | None = None,
        revision: str | None = None,
    ) -> dict[str, Any]:
        """Upload a gzipped tar archive; its ``id`` goes in ``workspaceId``.

        The body is the raw archive — not multipart. Uploading identical bytes
        twice returns the *pre-existing* record rather than creating a second
        one, so do not assume a fresh id per upload.
        """
        data = archive if isinstance(archive, bytes) else archive.read()
        headers = dict(self.connection.headers)
        headers.setdefault("User-Agent", "postgrip-agent-python")
        headers["Content-Type"] = "application/gzip"
        if self.connection.auth_token and "Authorization" not in headers:
            headers["Authorization"] = f"Bearer {self.connection.auth_token}"
        if repository_name:
            headers["X-PostGrip-Repository"] = repository_name
        if revision:
            headers["X-PostGrip-Revision"] = revision

        request = Request(
            self.connection.address + "/api/v1/workspaces",
            data=data,
            method="POST",
            headers=headers,
        )
        try:
            with urlopen(request, timeout=self.connection.timeout) as response:
                raw = response.read()
        except HTTPError as exc:
            raise RuntimeError(exc.read().decode() or str(exc)) from exc
        import json

        return json.loads(raw.decode()) if raw else {}

    def wait_until_running(
        self,
        sandbox_id: str,
        *,
        timeout: float = 120.0,
        poll_interval: float = 1.0,
    ) -> dict[str, Any]:
        """Poll until the sandbox is ready, fails, or the wait expires.

        Fails fast on ``failed`` with the sandbox's own message rather than
        burning the timeout, and the timeout error names the last observed
        state — "timed out" alone does not say whether the sandbox was still
        scheduling, mid-setup, or never picked up by an agent.
        """
        deadline = time.monotonic() + timeout
        last: dict[str, Any] = {}
        while True:
            record = self.get(sandbox_id)
            last = record
            if sandbox_is_ready(record):
                return record
            if record.get("observedState") == "failed":
                detail = (
                    record.get("failureMessage")
                    or record.get("failureCode")
                    or "no failure detail reported"
                )
                raise RuntimeError(f"postgrip-agent: sandbox {sandbox_id} failed: {detail}")
            if time.monotonic() >= deadline:
                raise TimeoutError(
                    f"postgrip-agent: sandbox {sandbox_id} was not running within {timeout}s "
                    f"(last observed state: {last.get('observedState', 'unknown')})"
                )
            time.sleep(poll_interval)

    # --- sessions ----------------------------------------------------------

    def open_session(
        self,
        sandbox_id: str,
        *,
        kind: str = "pty",
        command: list[str] | None = None,
        rows: int | None = None,
        columns: int | None = None,
        relay_base_url: str | None = None,
    ) -> "SandboxSession":
        """Create a session and dial the relay.

        Requires the optional ``websockets`` dependency::

            pip install "postgrip-agent[sandbox]"
        """
        session = self.create_session(
            sandbox_id, kind=kind, command=command, rows=rows, columns=columns
        )
        url = sandbox_relay_url(
            relay_base_url or self.connection.address, session["id"], session["ticket"]
        )
        headers = {}
        if self.connection.auth_token:
            # The ticket authorizes the session; the management credential
            # still authenticates the request. Both are required.
            headers["Authorization"] = f"Bearer {self.connection.auth_token}"
        return SandboxSession(url, headers)

    def exec(
        self,
        sandbox_id: str,
        command: list[str],
        *,
        stdin: bytes | None = None,
        relay_base_url: str | None = None,
    ) -> tuple[int | None, bytes]:
        """Run a command and return ``(exit_code, output)``.

        ``output`` carries stdout and stderr *interleaved* — the relay is a
        single stream, so they cannot be separated client-side. ``exit_code``
        is ``None`` when the close carried no code, meaning the transport ended
        rather than the process.
        """
        session = self.open_session(
            sandbox_id, kind="exec", command=command, relay_base_url=relay_base_url
        )
        with session:
            if stdin:
                session.send(stdin)
            chunks = list(session.read_all())
            return session.exit_code, b"".join(chunks)


class SandboxSession:
    """A live sandbox session: a raw bidirectional byte stream.

    There is no framing and no resize channel; rows and columns are fixed when
    the session is created.
    """

    def __init__(self, url: str, headers: dict[str, str]):
        try:
            from websockets.sync.client import connect  # type: ignore[import-not-found]
        except ImportError as exc:  # pragma: no cover - dependency guard
            raise RuntimeError(
                "postgrip-agent: sandbox sessions need the 'websockets' package; "
                'install with pip install "postgrip-agent[sandbox]"'
            ) from exc
        self._socket = connect(url, additional_headers=headers, max_size=None)
        self.exit_code: int | None = None

    def send(self, data: bytes | str) -> None:
        size = len(data.encode()) if isinstance(data, str) else len(data)
        if size > SANDBOX_RELAY_MAX_FRAME_BYTES:
            raise ValueError(
                f"postgrip-agent: sandbox write of {size} bytes exceeds the relay frame "
                f"limit ({SANDBOX_RELAY_MAX_FRAME_BYTES}); chunk your writes"
            )
        self._socket.send(data)

    def read_all(self) -> Iterable[bytes]:
        """Yield output until the session closes, recording the exit code."""
        from websockets.exceptions import ConnectionClosed  # type: ignore[import-not-found]

        try:
            for message in self._socket:
                yield message if isinstance(message, bytes) else message.encode()
        except ConnectionClosed as exc:
            self.exit_code = sandbox_exec_exit_code(exc.rcvd.code if exc.rcvd else 0)
        else:
            close_code = getattr(self._socket, "close_code", None)
            if close_code is not None:
                self.exit_code = sandbox_exec_exit_code(close_code)

    def close(self) -> None:
        self._socket.close()

    def __enter__(self) -> "SandboxSession":
        return self

    def __exit__(self, *exc_info: object) -> None:
        self.close()


def _sandbox_path(sandbox_id: str) -> str:
    if not sandbox_id:
        # Otherwise this builds /api/v1/sandboxes/ and hits the collection
        # route, which is a confusing 404 or worse a list.
        raise ValueError("postgrip-agent: sandbox id is required")
    return f"/api/v1/sandboxes/{quote(sandbox_id, safe='')}"
